### PR TITLE
Add basic IEEE-1588 PTP server/client

### PR DESCRIPTION
**Pull request by PetteriAimonen**
## Summary

This adds beginnings of a PTP (precision time protocol) daemon, which can work as server, client or both.

What works:
- Basic server & client operation
- Transmission and reception of announce, sync and follow-up
    
Still missing:
- SO_TIMINGS for getting more precise packet timestamps
- Implementation of delay_req and delay_resp packets
- Status and stop interfaces for the daemon

## Impact

Adds a new application `ptpd`.

## Testing

Tested against Linux ptpd2 version 2.3.1 using a custom STM32F4 board.

PC to NuttX sync (with debug messages enabled):

<pre>
linux$ sudo ptpd -i enp8s0 -V

nsh> ptpd eth0
Started the PTP daemon as PID=6
ptp_process_rx_packet: Got packet: 44 bytes
ptp_process_rx_packet: Got sync packet, seq 0
ptp_process_rx_packet: Got packet: 44 bytes
ptp_process_rx_packet: Got follow-up packet, seq 0
ptp_process_rx_packet: Got packet: 64 bytes
ptp_process_rx_packet: Got announce packet, seq 0
ptp_process_announce: Switching to better PTP time source
ptp_process_rx_packet: Got packet: 44 bytes
ptp_process_rx_packet: Got sync packet, seq 1
ptp_process_sync: Waiting for follow-up
ptp_process_rx_packet: Got packet: 44 bytes
ptp_process_rx_packet: Got follow-up packet, seq 1
ptp_update_local_clock: Jumped to timestamp 1695984277.193041402 s
ptp_process_rx_packet: Got packet: 44 bytes
ptp_process_rx_packet: Got sync packet, seq 2
ptp_process_sync: Waiting for follow-up
ptp_process_rx_packet: Got packet: 44 bytes
ptp_process_rx_packet: Got follow-up packet, seq 2
ptp_update_local_clock: Adjusting clock by 0.009991645 s
ptp_process_rx_packet: Got packet: 64 bytes
ptp_process_rx_packet: Got announce packet, seq 1
ptp_process_rx_packet: Got packet: 44 bytes
ptp_process_rx_packet: Got sync packet, seq 3
ptp_process_sync: Waiting for follow-up
ptp_process_rx_packet: Got packet: 44 bytes
ptp_process_rx_packet: Got follow-up packet, seq 3
ptp_update_local_clock: Adjusting clock by 0.000000000 s

nsh> date
Fri, Sep 29 10:44:56 2023
</pre>

NuttX to PC sync:

<pre>
linux$ sudo ptpd -i enp8s0 -V -s
2023-09-29 13:51:07.797108 ptpd2[88614].startup (info)      (___) Configuration OK
2023-09-29 13:51:07.797497 ptpd2[88614].startup (info)      (___) Successfully acquired lock on /var/run/ptpd2.lock
2023-09-29 13:51:07.797654 ptpd2[88614].startup (notice)    (___) PTPDv2 started successfully on wlp3s0 using "slaveonly" preset (PID 88614)
2023-09-29 13:51:07.797683 ptpd2[88614].startup (info)      (___) TimingService.PTP0: PTP service init
# Timestamp, State, Clock ID, One Way Delay, Offset From Master, Slave to Master, Master to Slave, Observed Drift, Last packet Received, One Way Delay Mean, One Way Delay Std Dev, Offset From Master Mean, Offset From Master Std Dev, Observed Drift Mean, Observed Drift Std Dev, raw delayMS, raw delaySM
2023-09-29 13:51:07.797717, init, 
2023-09-29 13:51:07.798156 ptpd2[88614].wlp3s0 (info)      (init) Observed_drift loaded from kernel: -7835 ppb
2023-09-29 13:51:07.898353 ptpd2[88614].wlp3s0 (notice)    (lstn_init) Now in state: PTP_LISTENING
2023-09-29 13:51:07.898401, lstn_init,  1 
2023-09-29 13:51:17.342671 ptpd2[88614].wlp3s0 (info)      (lstn_init) New best master selected: 120f03fffeffffff(unknown)/1
2023-09-29 13:51:26.044174 ptpd2[88614].wlp3s0 (notice)    (slv) Now in state: PTP_SLAVE, Best master: 120f03fffeffffff(unknown)/1
2023-09-29 13:51:26.044219, slv, 120f03fffeffffff(unknown)/1,  0.000000000,  0.000000000,  0.000000000,  0.000000000, -7834.701538086, I, 0.000000000, 0, 0.000000000, 0, 0, 0, -4.292860229,  0.000000000
2023-09-29 13:51:26.044372 ptpd2[88614].wlp3s0 (notice)    (slv) Received first Sync from Master
2023-09-29 13:51:26.044611, slv, 120f03fffeffffff(unknown)/1,  0.000000000, -4.208697754,  0.000000000, -4.208697754, -7834.701538086, S, 0.000000000, 0, 0.000000000, 0, 0, 0, -4.208697754,  0.000000000
2023-09-29 13:51:27.683261, slv, 120f03fffeffffff(unknown)/1,  0.000000000, -4.090707752,  0.000000000, -4.090707752, -7834.701538086, S, 0.000000000, 0, 0.000000000, 0, 0, 0, -4.090707752,  0.000000000
</pre>

### CODE REVIEW

File: examples/adjtime/adjtime_main.c [hunk 1]
```diff
@@ -148,7 +148,7 @@ int main(int argc, FAR char *argv[])
   parse_args(&delta, argc, argv);
 
   printf("Delta time is %ld seconds and %ld micro seconds.\n",
-         delta.tv_sec, delta.tv_usec);
+         (long)delta.tv_sec, delta.tv_usec);
```
Comment (pkarashchenko) [inline examples/adjtime/adjtime_main.c:151]: Maybe better cast to `intmax_t` and use `PRIdMAX` instead?

Comment (pkarashchenko) [reply examples/adjtime/adjtime_main.c:151]: Or choose printf specifier based of `CONFIG_SYSTEM_TIME64`

Comment (PetteriAimonen) [reply examples/adjtime/adjtime_main.c:151]: Technically yeah, but considering this is a delta time, it seems unlikely anyone would want to use the adjtime example program to change time by 70 years with rate adjustment.

The current implementation of `parse_args()` in the example limits values to `unsigned long`. I think that is a bit questionable, as the adjustment delta can be negative also. It does work with negative values currently, but that is probably more due to luck than design.

In any case, I consider the adjtime example quite low importance and I'm not interested in spending much time on it, just thought to fix a bug when I hit it.

Comment (pkarashchenko) [reply examples/adjtime/adjtime_main.c:151]: Ok. Let's keep it

File: netutils/ptpd/Kconfig [hunk 2]
```diff
@@ -0,0 +1,167 @@
+#
+# For a description of the syntax of this configuration file,
+# see the file kconfig-language.txt in the NuttX tools repository.
+#
+
+config NETUTILS_PTPD
+	bool "PTPD client/server"
+	default n
+	select NET_IPv4
+	select NET_IGMP
```
[the remark this answers is not in the document]
Comment (PetteriAimonen) [reply netutils/ptpd/Kconfig]: done

File: netutils/ptpd/ptpv2.h [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+/* FollowUp: actual timestamp of when sync message was sent */
+
+struct ptp_follow_up_s
+{
+  struct ptp_header_s header;
+  uint8_t origintimestamp[10];
+};
+
+#endif /* __APPS_NETUTILS_PTPD_PTPV2_H */
```
[the remark this answers is not in the document]
Comment (PetteriAimonen) [reply netutils/ptpd/ptpv2.h]: done

File: system/ptpd/Kconfig [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#
+# For a description of the syntax of this configuration file,
+# see the file kconfig-language.txt in the NuttX tools repository.
+#
+
+config SYSTEM_PTPD
+  tristate "PTP daemon commands"
```
Comment (pkarashchenko) [inline system/ptpd/Kconfig:7]: Please use TABs for indentation in Kconfig files

File: netutils/ptpd/ptpd.c [hunk 5]
```diff
@@ -0,0 +1,930 @@
...
+    {
+      return -1;
+    }
+  else
+    {
+      return pid;
+    }
+}
+
+/* TODO: Implement status and stop interfaces */
```
Comment (xiaoxiang781216) [inline netutils/ptpd/ptpd.c:930]: add blank line

File: netutils/ptpd/Makefile [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+############################################################################
+
+include $(APPDIR)/Make.defs
+
+# PTP server/client implementation
+
+ifeq ($(CONFIG_NET_UDP),y)
```
[the remark this answers is not in the document]
Comment (hartmannathan) [reply netutils/ptpd/Makefile:25]: Maybe intended check was CONFIG_SYSTEM_PTPD?

[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply netutils/ptpd/Makefile:25]: Don't need since the check is already done in Make.defs

[the remark this answers is not in the document]
Comment (PetteriAimonen) [reply netutils/ptpd/Makefile:25]: It was just copypaste from https://github.com/apache/nuttx-apps/blob/master/netutils/ntpclient/Makefile that I forgot to clean up, I'll fix that tomorrow.

[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply netutils/ptpd/Makefile:25]: Let's fix ntpclient to avoid the error happen again.

**Review by hartmannathan — APPROVED** [review]
LGTM

File: netutils/ptpd/ptpd.c [hunk 7]
```diff
@@ -0,0 +1,931 @@
...
+# define PTPD_POLL_INTERVAL CONFIG_NETUTILS_PTPD_TIMEOUT_MS
+#endif
+
+/* PTP debug messages are enabled by either CONFIG_DEBUG_NET_INFO
+ * or separately by CONFIG_NETUTILS_PTPD_DEBUG. This simplifies
+ * debugging without having excessive amount of logging from net.
+ */
+
+#ifdef CONFIG_NETUTILS_PTPD_DEBUG
+# ifndef CONFIG_DEBUG_NET_INFO
```
Comment (xiaoxiang781216) [inline netutils/ptpd/ptpd.c:137]: need two spaces

File: netutils/ptpd/ptpd.c [hunk 8]
```diff
@@ -0,0 +1,931 @@
...
+#   define ptpinfo _info
```
Comment (xiaoxiang781216) [inline netutils/ptpd/ptpd.c:138]: need four spaces

File: netutils/ptpd/ptpd.c [hunk 9]
```diff
@@ -0,0 +1,931 @@
...
+  } rxcmsg;
+
+  /* Buffered sync packet for two-step clock setting */
+
+  struct ptp_sync_s twostep_packet;
+  struct timespec twostep_rxtime;
+};
+
+#ifdef CONFIG_NETUTILS_PTPD_SERVER
+# define PTPD_POLL_INTERVAL CONFIG_NETUTILS_PTPD_SYNC_INTERVAL_MSEC
```
Comment (xiaoxiang781216) [inline netutils/ptpd/ptpd.c:126]: need two spaces

**Pull request merged**